15-connect-app-and-store/README.md

```diff
@@ -0,0 +1,527 @@
+
+```js
+import React, {PureComponent} from 'react';
+import PropTypes from 'prop-types';
+
+import {getEmails} from './action-reducers';
+
+export default class App extends PureComponent {
+  static propTypes = {
+    pollInterval: PropTypes.number
+  };
+
+  static defaultProps = {
+    // default the `pollInterval` prop to 2 secs when not specified
+    pollInterval: 2000
+  };
+
+  componentDidMount() {
+    // Retrieve emails from server once we know DOM exists
+    this._getUpdateEmails();
+
+    // Set up long-polling to continuously get new data
+    this._pollId = setInterval(
+      () => this._getUpdateEmails(),
+      this.props.pollInterval
+    );
+  }
+
+  componentWillUnmount() {
+    // Need to remember to clearInterval when the component gets
+    // removed from the DOM, otherwise the interval will keep going
+    // forever and leak memory
+    clearInterval(this._pollId);
+  }
+
+  _getUpdateEmails() {
+    return getEmails().then(emails => this.setState({emails}));
+  }
+
+  render() {
+    return (
+    );
+  }
+}
+```
+
+Now that we have separated out `<App/>` into two distinct components, lets start by setting up `<App/>` to create a `store` and have `<Provider/>` consume the store as the top level child. Then, we will return to `<Page/>` and complete it's refactor.
+
+Next, lets create our `store`.
+
+We are going to do this in `<App/>`'s [`constructor()`](https://facebook.github.io/react/docs/react-component.html#constructor). This should be added directly after `static defaultProps` as it is the first thing called when the component is evaluated.
+
+The only things we *need* when creating a store are: `createStore()` from `redux`, and our root reducer, which in this case is `emails` from [`reducers/index.js`]('.src/reducers/index.js'). Additionally, since we are using actions to make API calls which behave asynchronously, we also will need to import [`applyMiddleware()`](http://redux.js.org/docs/api/applyMiddleware.html) from `redux` and [`thunk`](https://github.com/gaearon/redux-thunk#whats-a-thunk) from `redux-thunk`.
+
+First, lets import the necessary modules:
+```js
+// App.js
+
+import {createStore, applyMiddleware} from 'redux';
+import thunk from 'redux-thunk';
+import {Provider} from 'react-redux';
+
+import {emails} from './reducers';
+```
+Next, lets add a `constructor()` in `<App />` which will handle creating the store.
+
+```js
+// App.js
+
+constructor(props) {
+  //It is always necessary to call super(props)
+  //whenever adding functionality to the constructor
+  super(props);
+
+  this._store = createStore(
+    emails,
+    applyMiddleware(thunk)
+  );
+}
+```
+We have a `store`! Now, that we do, we can modify our `render()` function so it returns `<Provider />` as the top level component, with `<Page />` as its child.
+
+```js
+//App.js
+
+render() {
+  <Provider store={this._store}>
+    <Page />
+  </Provider>
+}
+```
+
+After these changes our App.js should look something like:
```

Comment:
When pasting code, I try to only show the changes and use `// comments` for the stuff that remains the same.

See: https://github.com/benmvp/react-workshop/blame/master/09-mark-unread/README.md#L65